src/Workspaces/Core/Portable/Indentation/AbstractIndentationService.Indenter.cs

```diff
@@ -71,48 +70,96 @@ public IndentationResult GetDesiredIndentation(FormattingOptions.IndentStyle ind
-                if (indentStyle == FormattingOptions.IndentStyle.Block)
+                // we'll either be after the token at the end of a line, or before a token.  We compute indentation
+                // based on the preceding token.  So if we're before a token, look back to the previous token to
+                // determine what our indentation is based off of.
+                if (token.SpanStart >= LineToBeIndented.Start)
                 {
-                    // If it's block indentation, then just base 
-                    return GetIndentationOfLine(previousNonWhitespaceOrPreprocessorLine);
+                    token = token.GetPreviousToken();
+
+                    // Skip past preceding blank tokens.  This can happen in VB for example where there can be
+                    // whitespace tokens in things like xml literals.  We want to get the first visible token that we
+                    // would actually anch would anchor indentation off of.
+                    while (token != default && string.IsNullOrWhiteSpace(token.ToString()))
+                        token = token.GetPreviousToken();
                 }
 
-                Debug.Assert(indentStyle == FormattingOptions.IndentStyle.Smart);
+                if (token == default)
+                    return null;
+
+                return token;
+            }
 
-                // Because we know that previousLine is not-whitespace, we know that we should be
-                // able to get the last non-whitespace position.
-                var lastNonWhitespacePosition = previousNonWhitespaceOrPreprocessorLine.GetLastNonWhitespacePosition().Value;
+            private IndentationResult GetDesiredBlockIndentation()
+            {
+                // Block indentation is simple, we keep walking back lines until we find a line with any sort of
+                // text on it.  We then set our indentation to whatever the indentation of that line was.
+                for (var currentLine = this.LineToBeIndented.LineNumber - 1; currentLine >= 0; currentLine--)
+                {
+                    var line = this.Document.Text.Lines[currentLine];
+                    var offset = line.GetFirstNonWhitespaceOffset();
+                    if (offset == null)
+                        continue;
 
-                var token = Root.FindToken(lastNonWhitespacePosition);
-                Debug.Assert(token.RawKind != 0, "FindToken should always return a valid token");
+                    // Found the previous non-blank line.  indent to the same level that it is at
+                    return new IndentationResult(basePosition: line.Start + offset.Value, offset: 0);
+                }
 
-                return _service.GetDesiredIndentationWorker(
-                    this, token, previousNonWhitespaceOrPreprocessorLine, lastNonWhitespacePosition);
+                // Couldn't find a previous non-blank line.  Don't indent at all.
+                return default;
```

Comment:
Kinda weird to me this isn't a nullable return, as this is actually saying to align to the start of the file...

Reply:
changed to NRT.